Aquilon installation: document a VirtualEnv-based installation

jouvin opened the pull request:
Contribution from LAL workshop! Documentation of a VirtualEnv-based Aquilon installation, based on LAL experience.

Also fixes https://github.com/quattor/aquilon/issues/28

jouvin commented:
This PR is ready for review by an Aquilon expert! It should work both with the RPM-based approach documented so far and with the VirtualEnv-based approach introduced here.

ned21 commented on _aquilon/00-install.md:
```diff
@@ -201,14 +307,16 @@ To create the Aquilon database:
 
 To test the broker, run `/opt/aquilon/tests/dev_aqd.sh` that must run successfully if your
-installation is correct. Before running this script, you need to define `AQDCONF` environment
+installation is c orrect. Before running this script, you need to define `AQDCONF` environment
```
It looks like an extra space has been introduced here?

jouvin commented:
@ned21  the fact you are are starting from something which is not CentOS-based may not necessarily be a major issue with the VirtualEnv-based approach as there is very little that depends on the OS (git, make...) and you can probably easily add them even though this is a Debian-based container...

jouvin commented:
@urbonegi I'd be happy to get your feedback on this updated installation. I think it is much better than the previous version anyway... One recent change is the recommendation to start `git-daemon` from the broker. By default this is disabled but I am not sure why. It looks easier than configuring it with `inetd` on CentOS 7 due to problems in `ncm-systemd` (but it is pretty easier to do it manually at the price of duplicating part of the `/etc/aqd.conf` contents in a `/etc/sysconfig` file (or in the script to start the `git-daemon` if you accept to hardcode the values in it).

jouvin commented:
BTW, failing test is caused by the spellchecker which is not aware of many acronyms and usual words in our jargon... this is a known problem. @jrha : what about disabling this part of the check (or ensure that it return a success in any circumstance) or run it as a separate check so that there is no ambiguity about the reason of the failure?

ned21 commented on _aquilon/00-install.md:
```diff
@@ -201,6 +307,8 @@ To create the Aquilon database:
+# kinit user must be either `aquilon` if you use a standalone Kerberos server or
+# a valid principal if you have a central infrastructure (including Active Directory)
 kinit        # Enter the password you set previously
 /opt/aquilon/tests/aqdb/build_db.py
```
The `su - aquilon` command is not required if running within a virtualenv.

jouvin replied:
I don't understand why... If running a virtualenv as root, you remain root... The whole point is not to be root. But it should be mentioned to activate the virtualenv for this user too...

jouvin commented on _aquilon/00-install.md:
```diff
@@ -10,68 +10,139 @@ redirect_from:
+  gcc git git-daemon java-1.8.0-openjdk-devel libxslt libxml2 make panc \
+  knc krb5-workstation \
+  protobuf-compiler protobuf-devel \
+  python-devel python-setuptools
```
No I think this is a mistake done when releasing the new pan compiler. Probably missing in the release documentation...

jouvin commented on _aquilon/00-install.md:
````diff
@@ -10,68 +10,139 @@ redirect_from:
+In addition install the `sqlalchemy` module from PyPI (the EPEL version is too old):
 
 ```bash
 pip install sqlalchemy
````
@ned21 thanks for spotting this. It is obviously missing.  This is only in the case of a virtualenv that you don't need this RPM... but in this case `python-virtualenv`  was missing in the documentation... Fixing it.

jouvin commented on _aquilon/00-install.md:
````diff
@@ -10,68 +10,139 @@ redirect_from:
-* protobuf-compiler
-* protobuf-python
+    ```bash
+    virtualenv --prompt="(aquilon) " ~/aquilon-venv
````
Fixed.

jouvin commented on _aquilon/00-install.md:
```diff
@@ -10,68 +10,139 @@ redirect_from:
 
-Protocols:
+
+## Building Aquilon protocols
```
I'll do it.

jouvin commented on _aquilon/00-install.md:
````diff
@@ -10,68 +10,139 @@ redirect_from:
-make INSTALLDIR=/usr/local/lib/aquilon/protocols PROTOC=protoc install
-```
-Aquilon itself:
+## Getting Aquilon sources
````
I was surpised too (this is not from me). I'll remove them!

ned21 commented on _aquilon/00-install.md:
````diff
@@ -10,100 +10,154 @@ redirect_from:
+  protobuf-python python-pip
 ```
 
-### Contributing
````
I don't see this info anywhere else on https://www.quattor.org/documentation/ so I think it needs to remain for now or it won't be available anywhere - perhaps it could just be moved to the end as an Appendix?

jouvin replied:
This is what I did already in my last update to the PR yesterday... I should have said it explicitly in addition to mentioning in the commit message!

urbonegi commented on _aquilon/00-install.md:
```diff
@@ -158,8 +185,11 @@ changes are:
 * Comment out every else in the `tool_locations` section that point to a path
```
Actually we have created new config files for the community which would be even easier to use for Aquilon broker startup outside MS. The new config file is called aqd.conf.noms. This will just have to be moved to /etc/aqd.conf and amended if needed. I will push out the latest code to the aqd github repo today.

urbonegi replied:
Just updated the Aquilon broker code. This is the new conf file to use for outside the ms:
https://github.com/quattor/aquilon/blob/upstream/etc/aqd.conf.noms#L28

jouvin replied:
@urbonegi I just add a look to the new `aqd.conf.noms` file. I appreciate that it is much simpler but I'm afraid it is a little bit too simple in that it doesn't document anymore all the options related to directory configuration, meaning that you have to adhere to the default convention or you are lost... If you agree, I may propose a new version of the `noms` file readding what I think is useful in the previous `aqd.conf`  template.

BTW, the doc mentions (I kept it!) that you need to wipe the contents of `aqd.conf.defaults` which always looks strange to me... Is it still needed or is it no longer necessary?

urbonegi replied:
Of course, please do submit the PR for better version of aqd.conf.noms!! This is just example config for outside ms. I just added the few options that were documented in the install guide.
As for the wipe contents of the defaults - no that is not needed anymore - that was the whole  point of refactor i did :)
https://github.com/quattor/aquilon/commit/59ac3f19ee9c8ee54bb481cd83aeb341a2fb6422

ned21 commented:
I'm fine with this, anyone else have comments before we merge this?

jouvin commented:
@ned21 @urbonegi have you checked the new Contributing page. In particular the bits related to `Change-Id`. I am not sure I rephrased it correctly... and I realized that I should adhere to it for my own contributions!

jouvin commented:
@ned21 I realized that I forgot to add the file to this PR, sorry! Fixing this!

jouvin commented:
Everything should be there now! It would be good that one of the Aquilon expert takes a look at the Upgrade section added, in particular regarding schema upgrades.

jouvin commented:
The PR should be close to the perfection!! I started to add instructions for using Postgres as the db backend. I still have to write the § on the configuration steps, I hope to do it tomorrow... but all the other parts should be really final now.

stdweird commented on _aquilon/00-install.md:
```diff
@@ -133,6 +153,9 @@ you actually want to use:
 
+## Configuring Database
+
+This step applies only for installation using Postgres as their database backend (recommended, see above).
```
why not make a template and let quattor install the packages and configure the database (and aquilon itself)?

jouvin replied:
Because Aquilon may be installed before you have Quattor installed! When you already have SCDB running, you could imagine to do that: this is what we did in fact and I'm ready to share these templates if it helps. But this is probably outside the scope of this documentation.

stdweird commented:
@jouvin the only things you need are a template with spma (userpkgs enabled), list of repos to use (i think userpkgs enabled allows you to add repos without controlling them all); list of packages to install, some postgres configuration and some dirperm configuration.

make dummy object template, compile it, and bootstrap the node:
- git clone repo_with_profile
- sudo yum install -y epel-release
- rpm -Uvh http://yum.quattor.org/17.12.0/quattor-yum-repo-17.12.0-1.noarch.rpm
- yum install ncm-spma panc -y
- panc magic_template.pan
- edit /etc/ccm.conf, use `file://path/to /magic_template.xml`
- ccm-initialise
- ccm-fetch
- ncm-ncd --co -all

et voila (you can script this whole bootstrap process, so only 2 or 3 commands are required).

if you insist on documenting the individual steps, add annotations to the templates as usual.

best of all, you can edit the template while you are figuring out the steps; and use these same templates for full quattor managed box.

we should bootstrap quattor with quattor

jouvin commented:
@stdweird you are welcome to contribute this later, once this PR is merged! For the time being, I prefer to keep this approach. In particular as I want to document the VirtualEnv approach and be sure that there is no chicken&egg problem... I've been fighting too long with incomplete instructions, my priority is to get them complete! After we'll refine them.

jouvin commented:
If somebody has time to do a final review, I'd be interested to see this page merged to make my life easier (not a very strong argument I admit!) as I currently have to deal with 2 // PRs corresponding to the 2 different sections of the Aquilon documentation, going back and forth between them. I'm convinced the current content, if not completely accurate, is already a very significant improvement over the previous one that I wrote 18 months ago, in Annecy.

jouvin commented:
Ping!! I think it would really make sense to get this one merged. This will not prevent improving it again in a later PR but would minimize conflicts with the other PRs related to Aquilon docs.

ned21 commented on _aquilon/zz-contributing.md:
```diff
@@ -0,0 +1,38 @@
+for commit messages.
+
+[gerrit-commit-hook]: https://gerrit.googlesource.com/gerrit/+/master/resources/com/google/gerrit/server/tools/root/hooks/commit-msg
+{:target="_gerrit_commit_hook"}
```
This looks funny in the markdown viewer - does it show up OK when using the real website generator?

jouvin replied:
@ned21 Clearly, this syntax requires Jekyll (kramdown if I'm right) and is not properly supported by the markdown viewer integrated in GitHub web interface. I strongly advise you to install Jekyll on your local machine to be able to check the changes. This is (almost!) as simple as installing, Ruby, Ruby-gems and bundler. Once this is done, executing the following command should allow you to see the website on `localhost:4000`:

```bash
bundle exec jekyll s
```

jouvin commented:
I rebased this PR on #242 and added to the dictionary the missing words/acronyms. Somebody should review that my additions are appropriate. I was surprised to have to add `unsynchronized` (and may be `unsynchronised` would be more appropriate?).

jrha commented on .travis-scripts/dictionary.txt:
```diff
@@ -37,18 +38,27 @@ Kerberos
+Postgres
+postgres
+PostgresQL
+postgresql
```
"PostgreSQL" is the correct name, the others are all wrong.

jrha commented on .travis-scripts/dictionary.txt:
```diff
@@ -59,14 +69,23 @@ reconnection
+scalability
 SELinux
+SQL
+sql
```
IMHO SQL should always be capitalised

jrha commented on .travis-scripts/dictionary.txt:
```diff
@@ -59,14 +69,23 @@ reconnection
 unconfigured
 urbonegi
+unsynchronized
+userid
```
I don't think this is a word, it should probably be written "user ID"

jrha commented on .travis-scripts/dictionary.txt:
```diff
@@ -37,18 +38,27 @@ Kerberos
+libyaml
 logstash
 Maven
+md
```
Where did this get flagged? It should be written as "markdown" in a text, or `md` in code.

jrha commented on _aquilon/00-install.md:
```diff
@@ -1,109 +1,226 @@
+  * libxslt-devel (and its dependencies)
+  * libyaml-devel
+  * postgresql-devel
+  * python-virtualenv
```
If you quote these package names as code in ``, they will be ignored by the spellchecker.

jrha commented:
@jouvin the base `aspell-en` dictionary is quite small, but "unsynchronized" is perfectly valid.

jouvin commented:
So, AFAIC, this is ready to be merged.

jouvin commented:
Any reason to delay the merge of this PR? I'm happy to fix anything that you'll still find inappropriate... Feel free to squash all commits in one when merging as this is really only one big rewrite!
